### PR TITLE
Blocks Anonymous

### PR DESCRIPTION
**Pull request by scopatz**
What started off as a brief hack to add distributed support to xonsh morphed into adding full anonymous and onymous block support to xonsh.  This is implemented via a syntax tree transformation that raises an error with relevant data attached to it if a context manager has a true-valued `__xonsh_block__` attribute.

This PR adds two new context managers, `Block` and `Functor`. Blocks simply collect the lines, without executing the code:  

```python
x = 1
with Block() as b:
    x += 42

x == 1
b.lines == ['    x += 42']
```

Functor, on the other hand, uses the lines to create a function object that may be called later:

```python
x = [1]
with Functor() as f:
    x.append(5)
x.append(2)
f() or f.func()

x == [1, 2, 5]
```

These are meant as reusable components for building remote execution services, such as for SSH.  As a demonstration of this, there is now a distributed xontrib that you can use if you have started `dscheduler` and `dworker` in other process.  This xontrib adds a `DSubmitter` context manager and  a `dsubmit()` helper.  These can be used as follows:

```python
xontrib distributed

with dsubmit('127.0.0.1:8786', rtn='x') as d:
    x = $(echo dave is not here)
    x = x.strip()

# eventually
d.future.result() == 'dave is not here'
```

In any event, this got a little out of hand for what I originally intended.  However, I think it will pave the way for some pretty exciting features!

### CODE REVIEW

**Comment by mrocklin** [comment]
Um, whoa.  OK cool.  Will ponder.

**Comment by scopatz** [comment]
Thanks @mrocklin! Your feedback is definitely appreciated.

**Comment by Carreau** [comment]
@scopatz  : You are crazy.

**Comment by scopatz** [comment]
@Carreau I will take that as a xompliment, sir ;)

**Comment by Carreau** [comment]
> @Carreau I will take that as a xompliment, sir ;)

It is, it is.

**Comment by CJ-Wright** [comment]
@scopatz This is super exciting, although I need to play with it/think about it.

File: xontrib/distributed.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    from distributed import Executor
+    e = Executor(*a, **kw)
+    dsub = DSubmitter(e, args=args, kwargs=kwargs, rtn=rtn)
+    return dsub
+
+
+# some cleanup
+del dworker, Functor
+
```
Comment (adqm) [inline xontrib/distributed.py]: PEP8: blank line at end of file should be deleted.

File: xonsh/ast.py [hunk 2]
```diff
@@ -192,6 +235,79 @@ def is_in_scope(self, node):
...
+        not executed it will raise a XonshBlockError containing the required
+        information.
+        """
+        nwith = self._nwith  # the nesting level of the current with-statement
+        lineno = get_lineno(node)
+        col = get_col(node, 0)
+        # Add or discover target names
+        targets = set()
+        i = 0  # index of unassigned items
+        def make_next_target():
```
Comment (adqm) [inline xonsh/ast.py]: PEP8: there should be a blank line before this def

File: xonsh/ast.py [hunk 3]
```diff
@@ -192,6 +235,79 @@ def is_in_scope(self, node):
...
+                    targs = [make_next_target() for _ in item.context_expr.elts]
+                    optvars = Tuple(elts=targs, ctx=Store(), lineno=lineno,
+                                    col_offset=col)
+                else:
+                    optvars = make_next_target()
+                item.optional_vars = optvars
+            else:
+                targets.update(gather_names(item.optional_vars))
+        # Ok, now that targets have been found / created, make the actual check
+        # to see if we are in a non-executing block. This is equivalent to writing
```
Comment (adqm) [inline xonsh/ast.py]: PEP8: long line (but maybe we don't care since it's a comment?)

File: xonsh/ast.py [hunk 4]
```diff
@@ -192,6 +235,79 @@ def is_in_scope(self, node):
...
+        if len(tests) == 1:
+            test = tests[0]
+        else:
+            test = BoolOp(op=Or(), values=tests, lineno=lineno, col_offset=col)
+        ldx, udx = self._find_with_block_line_idx(node)
+        lines = [Str(s=s, lineno=lineno, col_offset=col)
+                 for s in self.lines[ldx:udx]]
+        check = If(test=test, body=[
+            Raise(exc=xonsh_call('XonshBlockError',
+                args=[List(elts=lines, ctx=Load(), lineno=lineno, col_offset=col),
```
Comment (adqm) [inline xonsh/ast.py]: PEP8: long line

Comment (adqm) [inline xonsh/ast.py]: Also, same line: indent too small for visual indent

File: xonsh/ast.py [hunk 5]
```diff
@@ -192,6 +235,79 @@ def is_in_scope(self, node):
...
+                      xonsh_call('globals', args=[], lineno=lineno, col=col),
+                      xonsh_call('locals', args=[], lineno=lineno, col=col)],
+                    lineno=lineno, col=col),
+                cause=None, lineno=lineno, col_offset=col)],
```
Comment (adqm) [inline xonsh/ast.py]: PEP8: indent too small for visual indent

File: xonsh/ast.py [hunk 6]
```diff
@@ -192,6 +235,79 @@ def is_in_scope(self, node):
...
+            orelse=[], lineno=lineno, col_offset=col)
+        node.body.insert(0, check)
+
+    def _find_with_block_line_idx(self, node):
+        ldx = min_line(node.body[0]) - 1
+        udx = max_line(node.body[-1])
+        # now check if parsable, or add lines until it is or we run out of lines
+        nlines = len(self.lines)
+        lines = 'with __xonsh_dummy__:\n' + '\n'.join(self.lines[ldx:udx]) + '\n'
```
Comment (adqm) [inline xonsh/ast.py]: PEP8: _just barely_ too long (81 chars)

File: xonsh/ast.py [hunk 7]
```diff
@@ -365,4 +484,17 @@ def pdump(s, **kwargs):
...
 
-
+def pprint(s, *, sep=None, end=None, file=None, flush=False, **kwargs):
+    """Performs a pretty print of the AST nodes."""
+    print(pdump(s, **kwargs), sep=sep, end=end, file=file, flush=flush)
+
+#
+# Private helpers
+#
+def _getblockattr(name, lineno, col):
```
Comment (adqm) [inline xonsh/ast.py]: PEP8: need another blank line

**Comment by adqm** [comment]
This is **_really_** cool!  I am happy to merge in (sorry it took so long to get to it), but I had a couple of questions:

* Are `Block`s good for anything on their own  (i.e., could we `exec` them in a given environment, or something like that?)?  I mean, they collect the lines, but it's not clear to me whether anything can actually be done with that (aside from the Functor awesomeness).  Presumably, it would be possible to add a hook to `xonsh.execer.Execer.exec` for executing a `Block` in a given environment?  I guess we would have to remove the first layer of indentation, but you're already doing something like that in the `Functor` class, so that seems doable.
* The `dscheduler` stuff is really neat!  Right now it relies on the remote `dworker`s having been started from within xonsh (with the `distributed` xontrib enabled).  Is there any way we could sneak this so that it would work for an arbitrary `dworker`?  My guess is no (and it's certainly not needed), but I figured I'd ask.  I got `NameError: name '__xonsh_subproc_captured_stdout__' is not defined` running with normal `dworker`s, which makes sense (I assume that that lgetting those variables was the reason for overriding `dworker` anyway?).

Also a few little PEP8 things, and this needs a (trivial) merge from master.  I'm happy to clean those up if you want.

**Comment by scopatz** [comment]
> This is really cool! I am happy to merge in (sorry it took so long to get to it), but I had a couple of questions:

No worries!  I think think that this is really going to up our game with respect to what other shells can do.

> Are Blocks good for anything on their own (i.e., could we exec them in a given environment, or something like that?)? I mean, they collect the lines, but it's not clear to me whether anything can actually be done with that (aside from the Functor awesomeness). Presumably, it would be possible to add a hook to xonsh.execer.Execer.exec for executing a Block in a given environment? I guess we would have to remove the first layer of indentation, but you're already doing something like that in the Functor class, so that seems doable.

Ahh, yeah, so blocks are meant as a more primitive tool.  Though presemuably, one could have an ExecutingBlock, easily by doing removing the leading whitespace (which sadly is a little more involved than just lstrip()).

I would say that the Functor awesome is only the begining and that the distinction here is done on purpose. For example, an SSH remote executor would presumably subclass Block, not Functor.  At least this is how I designed it.  

Another thing that one could potentially do would be to throw out the lines and simply use the Block to analyse the locals and globals.  I can't think of a specific use case, but....

Maybe one could write a (dynamic?) static analysis tool from Block, perhaps have a Block that is its own syntax tree transformer / code rewriter. There are a bunch of wacky/potentially useful ideas that all stem from having the source code.  This seemed like the write abstraction.

> The dscheduler stuff is really neat! Right now it relies on the remote dworkers having been started from within xonsh (with the distributed xontrib enabled). Is there any way we could sneak this so that it would work for an arbitrary dworker? My guess is no (and it's certainly not needed), but I figured I'd ask. I got NameError: name '__xonsh_subproc_captured_stdout__' is not defined running with normal dworkers, which makes sense (I assume that that lgetting those variables was the reason for overriding dworker anyway?).

Actually, this could be made to work using a normal dworker running from a Python  process, rather than a xonsh process.  *However* it would be super annoying to write.  What we would have to submit to the Executor in the DSubmitter would need to be a callable that knows how to pickle and unpickle all of the xonsh builtins on getstate/setstate calls, and send *that* over the wire.  We'd also have to figure out how to make sure that if a worker unpickled many callers, that the builtin state didn't overwrite each other, which probably means hiding the builtins in a closure.  

This is definitely the "Let's just run it in xonsh, problem solved!" implementation :)

> Also a few little PEP8 things, and this needs a (trivial) merge from master. I'm happy to clean those up if you want.

No worries! I fixed them when I merged from master.

**Comment by adqm** [comment]
Alright, cool.  This all looks ~~good~~ _awesome_ to me, then.  Any reason to wait on merging this in?

**Comment by melund** [comment]
I am also super excited about this. The possibilities of `Block`s really blows my mind. I can't beleive that anyone hasn't proposed this as a built-in python feature (PEP?).  

The first thin that came to my mind was an `ExecuteInCondaEnvBlock` to run code in specific conda environments. Something like what @pelson did with `conda-execute` but without the need for a stand-alone script. We could make the distinction between  xonsh and pure python blocks. Where pure python blocks could run in any environment.

**Comment by scopatz** [comment]
> Any reason to wait on merging this in?

No reason from my end!  Merge away!

> I am also super excited about this. The possibilities of Blocks really blows my mind. I can't beleive that anyone hasn't proposed this as a built-in python feature (PEP?).

People have, there are at least 4 related PEPs about Blocks, and what ultimately happened was the with-statement / context manager.  I think that the ones that failed didn't succeed because 

1. they proposed a new syntax, 
2. they were a form of "non-local flow control" (ie macros)
3. they where trying to be too much like Ruby, and truly anonymous (ie the new syntax).  

I think that this particular solution is unique though. I am pretty excited about it, actually.  I think it might be worth proposing as a PEP myself. But it is useful to have a working implementation (ie xonsh) and I am planning on talking to some people I know who have had successful PEPs to see how much pain it would be :).

> The first thin that came to my mind was an ExecuteInCondaEnvBlock to run code in specific conda environments. Something like what @pelson did with conda-execute but without the need for a stand-alone script. We could make the distinction between xonsh and pure python blocks. Where pure python blocks could run in any environment.

Yeah, for sure, that distinction would be nice to have, potentially!

**Comment by adqm** [comment]
> No reason from my end! Merge away!

Onward and upward!

**Comment by scopatz** [comment]
Woohoo!

**Comment by pelson** [comment]
:heart: - I want this in Python. Please do PEP it! :+1: :+1:

**Comment by scopatz** [comment]
With two thumbs up from @pelson, how can I say no :).  In the meantime though you can use xonsh if you want. Maybe too heavyweight, but just throwing that out there!

**Comment by pelson** [comment]
> With two thumbs up from @pelson, how can I say no

😉 

So I was playing with this a little and wanted to be able to get the contents of Block without transformation but it looks like parsing takes place before XonshBlockError is raised. Possible, or has the execer parsing ship sailed by that point ⛵ ? I completely accept that doing so couple conceptually break the statement "All of the context sensitive parsing occurs as an AST transformation prior to any code is executed".

Essentially, I'd love to have a context manager which allowed me to:

```
with you_know_this_isnt_valid_syntax() as b:
     what the!?

b.lines == ['    what the!?']
```

and yes, I'm completely aware that what I'm asking for looks very similar to a string literal... 😄

**Comment by pelson** [comment]
Also spotted some behaviour that may be unexpected:

```
$ cat no_parse.xsh 
from xonsh.contexts import Block

with Block() as b:
    ls -l

print(b.lines)
```

```
$ xonsh no_parse.xsh 
['from xonsh.contexts import Block', '', '', 'with Block() as b:', '    ls -l', '', '', 'print(b.lines)']
```

But the following code:
```
from xonsh.contexts import Block

with Block() as b:
    a = 2

print(b.lines)
```

produces the expected list:

```['    a = 2']```

**Comment by scopatz** [comment]
Hi @pelson, first off, thanks for reporting that error. I'll get right on that.

Secondly, I think what you are actually looking for is a markup language :).  The problem runtime languages face in this instance is how do you flag to the interpreter to stop parsing with out extra syntax.  That is, if all you see is the expr string `you_know_this_isnt_valid_syntax()`, you actually have no idea whether that is going to return an object that tell you to stop parsing or not until you have actual called that function and created that object.

Since language go tokenizer -> lexer -> parser -> compiler -> exec, by the time you have the object in exec, the interpreter is well past the ability to go back and stop parsing.  This is why there is syntax for simply collecting bytes, as you said strings :).  Couple that with the fact that most languages don't want to add two ways of doing the same thing (unless it is C++) and you'd be hard pressed to get this behaviour anywhere.

Markup languages can get away with this because they don't have a notion of execution at all, of course.

I'll try to think up a trick that can do this by suppressing SyntaxErrors or something, but I am not 100% confident that is possible. And even if I could, I would highly doubt that a PEP to that effect would be accepted. One could just use strings after all :)

**Comment by pelson** [comment]
Thanks @scopatz. I agree with all you have said. Like you point out, I'm really talking about markup to break out of the parser - it equates pretty much to the magics concept in IPython and I'm effectively looking for a way to achieve that with pythonic syntax. For example:

```
%%cython
def sum(int a, int b):
    cdef int s = a+b
    return s
```

becomes

```
with cython_magic as m:
    def sum(int a, int b):
        cdef int s = a+b
        return s

m.sum(1, 2) == 3
```

Frankly, I completely accept the argument that it is pretty ill-conceived (and in this case doesn't even pose any benefit) and that I should just use a litteral, but was curious enough about your `Blocks` implementation to try 😉 .

**Comment by scopatz** [comment]
Yeah, I hear you :) Cython has something that is effectively like this with its `nogil` functionality.  We could definitiely do something similar, but even better would be for this ability to be subclassed.  I'll keep thinking about this.  There might be an option that is reasonable.

**Comment by scopatz** [comment]
Also, a potential fix is in for the issue that you noticed in #1199

**Pull request merged**